### PR TITLE
fix(database_observability.postgres): Correctly handle dsn with unix sockets

### PR DESCRIPTION
**Pull request by k3-cat**
### Brief description of Pull Request

Fixes `dsnParseRegex` so that `replaceDatabaseNameInDSN` can properly handle dsn with unix sockets, instead of treating the final part of sockets path as dbname.

### Pull Request Details

<details open>
  <summary>Current behaviour</summary>

  image

  - Thus, alloy will try to connect `/run/{dbname}/.s.PGSQL.5432` and produce error: `dial unix /run/{dbname}/.s.PGSQL.5432: connect: no such file or directory`

</details>

<details open>
  <summary>Fixed behaviour</summary>

  image

</details>

Although simply replace the `.+` in matching group 1 with `.*?` or `[^?]*`  can solve this issue as well, the current approach provides better robustness for future changes and edge cases (eg. unescaped spaces present in dsn).

### Issue(s) fixed by this Pull Request

nil

### Notes to the Reviewer

- matching group 3 is also rewired for better readability (as `.*` would match `\??` anyway) and robustness under edge cases
- `$` in matching group 3 is dropped intentionally, as this can generally improve the performance sightly 

### PR Checklist

- [ ] Documentation added (maybe not applicable?)
- [x] Tests updated
- [ ] Config converters updated (not applicable)

### CODE REVIEW

**Comment by fridgepoet** [comment]
Thank you for raising this issue and working on it!

>Although simply replace the .+ in matching group 1 

How about replacing it with `[^/]*`? 

I'm inclined to go with a more conservative approach that will support your unix sockets use case without changes to the current behavior.

**Comment by k3-cat** [comment]
yes yes, that would also work

File: internal/component/database_observability/postgres/collector/dsn.go [hunk 1]
```diff
@@ -8,7 +8,7 @@ import (
 
 type databaseConnectionFactory func(dsn string) (*sql.DB, error)
 
-var dsnParseRegex = regexp.MustCompile(`^(\w+:\/\/.+\/)(?<dbname>[\w\-_\$]+)(\??.*$)`)
+var dsnParseRegex = regexp.MustCompile(`^(\w+:\/\/[^\/]*\/)(?<dbname>[\w\-_\$]+)((?:\?\S*)?)`)
```
Comment (fridgepoet) [inline internal/component/database_observability/postgres/collector/dsn.go:11]:
```suggestion
var dsnParseRegex = regexp.MustCompile(`^(\w+:\/\/[^\/]*\/)(?<dbname>[\w\-_\$]+)(\??.*$)`)
```

Same here, let's just leave the old, less restrictive regex here.

File: internal/component/database_observability/postgres/collector/dsn_test.go [hunk 2]
```diff
@@ -39,17 +39,35 @@ func TestReplaceDatabaseNameInDSN(t *testing.T) {
 			expected:  "postgres://postgres:password@localhost:5432/testdb",
 		},
 		{
-			name:      "database name appears in password",
+			name:      "problematic case - database name appears in password",
```
Comment (fridgepoet) [inline internal/component/database_observability/postgres/collector/dsn_test.go:42]:
```suggestion
			name:      "database name appears in password",
```

Nit: Let's just keep this problematic part out of all the titles.

File: internal/component/database_observability/postgres/collector/dsn_test.go [hunk 3]
```diff
@@ -39,17 +39,35 @@ func TestReplaceDatabaseNameInDSN(t *testing.T) {
...
 			dsn:       "postgres://user:mydb123@localhost:5432/mydb",
 			newDBName: "newdb",
 			expected:  "postgres://user:mydb123@localhost:5432/newdb",
 		},
 		{
-			name:      "database name with special characters",
+			name:      "problematic case - database name with special characters",
```
Comment (fridgepoet) [inline internal/component/database_observability/postgres/collector/dsn_test.go:48]:
```suggestion
			name:      "database name with special characters",
```

Nit: Let's just keep this problematic part out of all the titles.

File: internal/component/database_observability/postgres/collector/dsn_test.go [hunk 4]
```diff
@@ -62,6 +80,18 @@ func TestReplaceDatabaseNameInDSN(t *testing.T) {
...
+			dsn:         "postgres://user:pass @localhost:5432/",
+			newDBName:   "newdb",
+			expectError: true,
+		},
+		{
+			name:        "DSN with space in query parameters",
+			dsn:         "postgres://user:pass@localhost:5432/mydb?sslmode=disable &connect_timeout=10",
+			newDBName:   "newdb",
+			expectError: true,
+		},
```
Comment (fridgepoet) [inline internal/component/database_observability/postgres/collector/dsn_test.go:94]:
```suggestion
```

Would suggest deleting this, we don't need to support this behavior in this PR.

File: internal/component/database_observability/postgres/collector/dsn_test.go [hunk 5]
```diff
@@ -39,17 +39,35 @@ func TestReplaceDatabaseNameInDSN(t *testing.T) {
...
+			dsn:       "postgres://user:@/mydb?host=/run/postgresql",
+			newDBName: "newdb",
+			expected:  "postgres://user:@/newdb?host=/run/postgresql",
+		},
+		{
+			name:      "unix socket problematic case - hostname and dbname with special characters",
+			dsn:       "postgres://user:pass@ex-amp_le.com:5432/my-db_test$1?host=/run/postgresql",
+			newDBName: "new_db",
+			expected:  "postgres://user:pass@ex-amp_le.com:5432/new_db?host=/run/postgresql",
+		},
```
Comment (fridgepoet) [inline internal/component/database_observability/postgres/collector/dsn_test.go:70]:
```suggestion
		{
			name:      "hostname with special characters",
			dsn:       "postgres://user:pass@ex-amp_le.com:5432/mydb",
			newDBName: "new_db",
			expected:  "postgres://user:pass@ex-amp_le.com:5432/new_db",
		},
```

Hmm here I was trying to understand what we are testing. Is it a hostname with special characters? 
I think the unix socket query parameter might not feel realistic here, so I removed here.

This is my suggestion, happy to discuss.

**Review by fridgepoet — COMMENTED** [review]
Thank you for having a look again!
I appreciate you raising the unix sockets issue and taking the time to add support for it here. I just added some more small suggestions and then we can try to merge.

P.S. I have merged main into your branch.

**Comment by fridgepoet** [comment]
Hi @k3-cat do you need any help getting this over the line? Feel free to let us know!

**Comment by fridgepoet** [comment]
Thank you so much @k3-cat for your contribution. I've implemented this in #6463 .
